### ISSUE TITLE
Problems happened in the make process

### ISSUE THREAD

**Issue by popoyatou** [iotk/iochibity-java#1]
Hello, I am trying to build the JNI layer, however some problems happened and I could not make it. 
I cloned the iochibity-java and use the official iotivity develop kit which I downloaded from the iotivity website. 
after entering "make TARGET_OS=linux ARCH=x86_64", here are some questions I have:
1. Which version of the iotivity should I use? I have tried with the iotivity-1.1.1, iotivity-1.2.0 and the latest code from the iotivity github, but with all the versions there are some problems happened, either no such file or the function is not defined.
2. Do I need to compile the iotivity code first using "scons TARGET_OS=android" or "scons TARGET_OS=linux TARGET_ARCH=x86_64" under the iotivity folder before build the JNI layer? Because with the latest code of iotivity, an error happened  regarding "iotivity_config.h" is missing, which seems to be a file that is generated during the scons process.

Thank you very much for your help!

Best Regards,
Jie

**Comment by popoyatou** [comment]
after compile first the iotivity code, the process got to continue. However, I am current stuck at the point for compiling the JniOcRepresentation.o. The process terminates with a not very clear notice: "make: **\* [JniOcRepresentation.o] Error 1".
There are several notes and warnings such as: 
"/usr/local/include/boost/variant/variant.hpp: In member function 'boost::detail::variant::invoke_visitor<Visitor>::result_type boost::detail::variant::invoke_visitor<Visitor>::internal_visit(T&, int) [with T = const std::vector<OCByteString>; Visitor = const JObjectConverter; boost::detail::variant::invoke_visitor<Visitor>::result_type = _jobject*]':
/usr/local/include/boost/variant/variant.hpp:979:5: warning: control reaches end of non-void function [-Wreturn-type]"  or 
"JniOcRepresentation.h:582:13: note:   no known conversion for argument 1 from 'std::vectorstd::vector<std::vector<OCByteString > >' to 'const std::vectorstd::vector<std::vector<OC::OCRepresentation > >&'" 

However, the only error I have seen is "/usr/local/include/boost/variant/variant.hpp:978:32: error: no match for call to '(const JObjectConverter) (std::vectorstd::vector<std::vector<OCByteString > >&)'" . I have tried with the boost of version 1.58.0, 1.60.0 and 1.61.0 but the same thing happens. I wonder do you have any clue about this? Thank you very much. 

Best Regards,
Jie

**Comment by mobileink** [comment]
On Fri, Nov 18, 2016 at 6:13 PM, popoyatou notifications@github.com wrote:

> after compile first the iotivity code, the process got to continue.
> However, I am current stuck at the point for compiling the
> JniOcRepresentation.o. The process terminates with a not very clear notice:
> "make: **\* [JniOcRepresentation.o] Error 1".
> There are several notes and warnings such as: "/usr/local/include/boost/variant/variant.hpp:
> In member function 'boost::detail::variant::invoke_visitor::result_type
> boost::detail::variant::invoke_visitor::internal_visit(T&, int) [with T =
> const std::vector; Visitor = const JObjectConverter;
> boost::detail::variant::invoke_visitor::result_type = _jobject*]':
> /usr/local/include/boost/variant/variant.hpp:979:5: warning: control
> reaches end of non-void function [-Wreturn-type]" or
> "JniOcRepresentation.h:582:13: note: no known conversion for argument 1
> from 'std::vectorstd::vector<std::vector<OCByteString > >' to 'const
> std::vectorstd::vector<std::vector<OC::OCRepresentation > >&'"
> However, the only error I have seen is "/usr/local/include/boost/variant/variant.hpp:978:32:
> error: no match for call to '(const JObjectConverter)
> (std::vectorstd::vector<std::vector<OCByteString > >&)'" . I have tried
> with the boost of version 1.58.0, 1.60.0 and 1.61.0 but the same thing
> happens. I wonder do you have any clue about this? Thank you very much.
> 
> Best Regards,
> Jie

Hi Jie,

If you can wait a day or two I've got new versions almost ready to push.
I'll try to reproduce your issue.

-gregg

**Comment by popoyatou** [comment]
Ok, thank you very much! If you need more detailed information for debugging, just let me know. :)

**Comment by mobileink** [comment]
On Fri, Nov 18, 2016 at 6:13 PM, popoyatou notifications@github.com wrote:

> after compile first the iotivity code, the process got to continue.
> However, I am current stuck at the point for compiling the
> JniOcRepresentation.o. The process terminates with a not very clear notice:
> "make: **\* [JniOcRepresentation.o] Error 1".
> There are several notes and warnings such as: "/usr/local/include/boost/variant/variant.hpp:
> In member function 'boost::detail::variant::invoke_visitor::result_type
> boost::detail::variant::invoke_visitor::internal_visit(T&, int) [with T =
> const std::vector; Visitor = const JObjectConverter;
> boost::detail::variant::invoke_visitor::result_type = _jobject*]':
> /usr/local/include/boost/variant/variant.hpp:979:5: warning: control
> reaches end of non-void function [-Wreturn-type]" or
> "JniOcRepresentation.h:582:13: note: no known conversion for argument 1
> from 'std::vectorstd::vector<std::vector<OCByteString > >' to 'const
> std::vectorstd::vector<std::vector<OC::OCRepresentation > >&'"
> However, the only error I have seen is "/usr/local/include/boost/variant/variant.hpp:978:32:
> error: no match for call to '(const JObjectConverter)
> (std::vectorstd::vector<std::vector<OCByteString > >&)'" . I have tried
> with the boost of version 1.58.0, 1.60.0 and 1.61.0 but the same thing
> happens. I wonder do you have any clue about this? Thank you very much.

FYI it looks like I'm getting the same error when I try to build
iochibity-java against 1.2.0.  I built 1.2.0 using just '$ scons', nothing
extra.  I think I have boost 1.54 installed. I'll look into it and get back
to you.

Note that iochibity-java will not compile against 1.1.1, since it depends
on some changes that did not make it in to 1.1.1.

-Gregg

**Comment by popoyatou** [comment]
So are you going to build the iochibity-java against 1.2.0 or the latest code from the iotivity github? (In case that the latest code has some changes on the basis of the 1.2.0).

**Comment by mobileink** [comment]
On Sat, Nov 19, 2016 at 2:28 PM, popoyatou notifications@github.com wrote:

> So are you going to build the iochibity-java against 1.2.0 or the latest
> code from the iotivity github? (In case that the latest code has some
> changes on the basis of the 1.2.0).

Sorry, I should have said the latest from branch "master" of the "iotivity"
repo.  The 1.2.0 tarball does not build for me!

-g

**Comment by popoyatou** [comment]
Hi, Gregg, 

How is the new version going? xD

**Comment by mobileink** [comment]
On Wed, Nov 23, 2016 at 7:41 AM, popoyatou notifications@github.com wrote:

> Hi, Gregg,
> 
> How is the new version going? xD

In progress.  I'm playing at a completely new Java API which is a minimal
wrapper on the C API.  Look for a message to the mailing list in the next
few days.

If you don't mind my asking, how do you expect to use this?  What platforms
are you interested in, what kind of functionality, etc.?  The new API will
be minimal, which opens the possibility of building services in Java (or
for that matter, any language that targets the JVM, Clojure in my case),
instead of C++.

Thanks,

Gregg

**Comment by popoyatou** [comment]
My objective is to build a small Java application project for the Web of Things. The project currently runs on my linux system (Linux Parallels-Virtual-Platform 4.4.0-38-generic #57-Ubuntu SMP x86_64 x86_64 x86_64 GNU/Linux) to get the resources, and do CRUDN operations on the resources. The intention is to map the resources to a semantic model (a semantic description of things) so that the resources can be exposed to the user and be controlled through the semantic model. I would like a Java API as for the semantic part, I work with Java. 
Basically the functionality I want is quite similar as the ones provided by the android-api of the Iotivity. I was trying to convert the aar file directly to a jar, but I found out that the .so files included in the aar file have dependencies on some .so files of android, so I could not make the conversion succeed. ( I am currently testing the whole idea on an android phone by using the android-api, and it would be great that when the Java API works, I can easily transplant my program from the phone to the computer. ）

**Comment by popoyatou** [comment]
Hi, Gregg, how are you? Have you already started making the new Java API?

**Comment by mobileink** [comment]
Hi Jie,

Sorry for the delay.  The new API is looking pretty good, IMHO.  I'm just now finishing up the GET functionality, after which the remaining methods should be easy.  No documentation yet, but you can take a look at the jni-c directory, especially the example subdir, to get an idea of what it looks like.

-Gregg

**Comment by mobileink** [comment]
FYI I just uploaded some updated documentation.
